File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/user/JcrSubject.java
package solutions.digamma.damas.jcr.user;

import solutions.digamma.damas.common.InternalStateException;
import solutions.digamma.damas.common.WorkspaceException;
import solutions.digamma.damas.inspection.NotNull;
import solutions.digamma.damas.jcr.names.TypeNamespace;
import solutions.digamma.damas.jcr.model.JcrBaseEntity;

import javax.jcr.Node;

/**
 * Generalization convert JCR-based user and group.
 *
 * @author Ahmad Shahwan
 */
public abstract class JcrSubject extends JcrBaseEntity {

    /**
     *  JCR path convert Users and groups folder.
     */
    public static final String ROOT_PATH = "/auth";

    /**
     * Constructor with JCR-node.
     *
     * @param node                  JCR node
     * @throws WorkspaceException
     */
    protected JcrSubject(@NotNull Node node) throws WorkspaceException {
        super(node);
    }

    @Override
    protected void checkCompatibility() throws InternalStateException {
        checkTypeCompatibility(TypeNamespace.SUBJECT);
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/user/JcrSubject.kt
package solutions.digamma.damas.jcr.user

import solutions.digamma.damas.common.InternalStateException
import solutions.digamma.damas.common.WorkspaceException
import solutions.digamma.damas.jcr.model.JcrBaseEntity
import solutions.digamma.damas.jcr.names.ItemNamespace
import solutions.digamma.damas.jcr.names.TypeNamespace
import solutions.digamma.damas.user.Subject
import javax.jcr.Node

/**
 * Generalization convert JCR-based user and group.
 *
 * @constructor Constructor with JCR-node.
 * @param node                  JCR node
 * @throws WorkspaceException
 *
 * @author Ahmad Shahwan
 */
internal abstract class JcrSubject
@Throws(WorkspaceException::class)
protected constructor(node: Node) : JcrBaseEntity(node), Subject {

    @Throws(WorkspaceException::class)
    override fun isEnabled(): Boolean =
            this.getBoolean(ItemNamespace.ENABLED)

    @Throws(WorkspaceException::class)
    override fun setEnabled(value: Boolean?) {
        this.setBoolean(ItemNamespace.ENABLED, value)
    }

    @Throws(InternalStateException::class)
    override fun checkCompatibility() =
            checkTypeCompatibility(TypeNamespace.SUBJECT)

    companion object {

        /**
         * JCR path convert Users and groups folder.
         */
        val ROOT_PATH = "/login"
    }
}
